### PR TITLE
storage: add metrics for replica queues

### LINKED ISSUE #9272: storage: replica queues need metrics
The replica queues need per-queue metrics:
- Number of replicas processed successfully
- Number of errors processing replicas
- Number of replicas in purgatory
- Replica processing time

Cc @cockroach/stability

### PR DESCRIPTION
**Pull request by spencerkimball**
In particular: total counts for successfully processed and process
failures, current length of pending queue, total nanoseconds of
replica processing time, and length of purgatory queue.

Fixes #9272

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9398)

### CODE REVIEW

**Comment by spencerkimball** [comment]
@cockroachdb/stability

File: storage/metrics.go [hunk 1]
```diff
@@ -343,6 +421,33 @@ func newStoreMetrics() *StoreMetrics {
 		raftRcvdMessages:          make(map[raftpb.MessageType]*metric.Counter, len(raftpb.MessageType_name)),
 
 		RaftEnqueuedPending: metric.NewGauge(metaRaftEnqueuedPending),
+
+		// Replica queue metrics.
+		GCQueueProcessed:          metric.NewCounter(metaGCQueueProcessed),
```
Comment (tschottdorf) [inline storage/metrics.go]: The GCQueue could track more interesting information. See the `GCInfo` struct:

```go
type GCInfo struct {
	// Now is the timestamp used for age computations.
	Now hlc.Timestamp
	// Policy is the policy used for this garbage collection cycle.
	Policy config.GCPolicy
	// Stats about the userspace key-values considered, namely the number of
	// keys with GC'able data, the number of "old" intents and the number of
	// associated distinct transactions.
	GCKeys, IntentsConsidered, IntentTxns int
	// TransactionSpanTotal is the total number of entries in the transaction span.
	TransactionSpanTotal int
	// Summary of transactions which were found GCable (assuming that
	// potentially necessary intent resolutions did not fail).
	TransactionSpanGCAborted, TransactionSpanGCCommitted, TransactionSpanGCPending int
	// AbortSpanTotal is the total number of transactions present in the abort cache.
	AbortSpanTotal int
	// AbortSpanConsidered is the number of abort cache entries old enough to be
	// considered for removal. An "entry" corresponds to one transaction;
	// more than one key-value pair may be associated with it.
	AbortSpanConsidered int
	// AbortSpanGCNum is the number of abort cache entries fit for removal (due
	// to their transactions having terminated).
	AbortSpanGCNum int
	// PushTxn is the total number of pushes attempted in this cycle.
	PushTxn int
	// ResolveTotal is the total number of attempted intent resolutions in
	// this cycle.
	ResolveTotal int
	// ResolveErrors is the number of successful intent resolutions.
	ResolveSuccess int
	// Threshold is the computed expiration timestamp. Equal to `Now - Policy`.
	Threshold hlc.Timestamp
}
```

File: storage/queue.go [hunk 2]
```diff
@@ -147,6 +148,16 @@ type queueConfig struct {
 	acceptsUnsplitRanges bool
 	// processTimeout is the timeout for processing a replica.
 	processTimeout time.Duration
+	// processedMetric is a count of total replicas processed successfully.
```
Comment (tschottdorf) [inline storage/queue.go]: `s/count of total/total count of/g` throughout?

File: storage/queue.go [hunk 3]
```diff
@@ -511,11 +521,19 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) error {
 
 	log.VEventf(3, bq.ctx, "processing")
 	start := timeutil.Now()
-	if err := bq.impl.process(ctx, clock.Now(), repl, cfg); err != nil {
+	err := bq.impl.process(ctx, clock.Now(), repl, cfg)
+	duration := timeutil.Since(start)
+	log.VEventf(2, bq.ctx, "done (err=%s): %s", err, duration)
```
Comment (tschottdorf) [inline storage/queue.go]: use `%v` for this error since `err == nil` is possible.

Comment (tschottdorf) [inline storage/queue.go]: also there's `log.ErrEvent{,f}` and we should probably get in the habit of using it whenever `err` or `pErr` is involved.

File: storage/queue.go [hunk 4]
```diff
@@ -526,6 +544,11 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) error {
 // mechanism for signaling re-processing of replicas held in
 // purgatory.
 func (bq *baseQueue) maybeAddToPurgatory(repl *Replica, err error, clock *hlc.Clock, stopper *stop.Stopper) {
+	// Increment failures metric.
+	if bq.failuresMetric != nil {
```
Comment (tschottdorf) [inline storage/queue.go]: Can we not make them never nil by creating dummy objects if none are provided?

File: storage/queue.go [hunk 5]
```diff
@@ -511,11 +521,19 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) error {
...
+	log.Trace(ctx, "done")
+	if bq.timeMetric != nil {
+		bq.timeMetric.Inc(duration.Nanoseconds())
+	}
+	if err != nil {
 		return err
 	}
-	log.VEventf(2, bq.ctx, "done: %s", timeutil.Since(start))
-	log.Trace(ctx, "done")
+	if bq.processedMetric != nil {
```
Comment (mberhault) [inline storage/queue.go]: wouldn't this make more sense next to `bq.failuresMetric`?

**Comment by tamird** [comment]
Reviewed 8 of 10 files at r1.
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/gc_queue.go, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRirtwhG4IzF58X9wjr:-KRirtwhG4IzF58X9wjs:bnuh9hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/gc_queue.go#L105)):*
> ```Go
> 		needsLease:           true,
> 		acceptsUnsplitRanges: false,
> 		processedMetric:      store.metrics.GCQueueProcessed,
> ```

to echo @petermattis from yesterday: this is Go, not obj-c, so I think we can omit the type from the name. Perhaps "processedReplicas"? ditto for the other fields.

---

*[storage/queue.go, line 528 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRisCwuTBsu02X9vxT_:-KRisCwuTBsu02X9vxTa:bdjjc85) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L528)):*
> ```Go
> 	log.VEventf(2, bq.ctx, "done (err=%s): %s", err, duration)
> 	log.Trace(ctx, "done")
> 	if bq.timeMetric != nil {
> ```

throughout: why are these nil checks necessary? if it's a matter of test construction, let's fix the tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

File: storage/queue_test.go [hunk 6]
```diff
@@ -626,6 +724,59 @@ func TestBaseQueueProcessTimeout(t *testing.T) {
...
+	processM, _, _, timeM, _ := addMetrics(&bq)
+	bq.Start(tc.clock, tc.stopper)
+	bq.MaybeAdd(r, hlc.ZeroTimestamp)
+
+	util.SucceedsSoon(t, func() error {
+		if v := processM.Count(); v != 1 {
+			return errors.Errorf("expected 1 processed replicas; got %d", v)
+		}
+		if v := timeM.Count(); v < (5 * time.Millisecond).Nanoseconds() {
+			return errors.Errorf("expected >= 5e6 ns in processing time; got %d", v)
```
Comment (tschottdorf) [inline storage/queue_test.go]: nit:

```go
if min, v := 5*time.Millisecond, timeM.Count(); v < min.Nanoseconds() {
  return errors.Errorf("expected >= %s ns in processing time; got %d", e, v)
}
```

**Comment by tamird** [comment]
Review status: 8 of 9 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/queue_test.go, line 740 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiu1QrOB-1DL4AmRWJ:-KRiu1QrOB-1DL4AmRWK:bd8dxq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue_test.go#L740)):*
> ```Go
> 
> func (pq *processTimeQueueImpl) process(
> 	ctx context.Context,
> ```

all these arguments can be named `_`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

**Review by petermattis — APPROVED** [review]
LGTM

**Comment by tamird** [comment]
Reviewed 2 of 10 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[storage/metrics.go, line 153 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjCHtRUb02fuSQQmwy:-KRjCHtSTPsJA_KSfBpd:b-2weso3) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L153)):*
> ```Go
> 	metaRaftEnqueuedPending = metric.Metadata{Name: "raft.enqueued.pending",
> 		Help: "Number of pending outgoing messages in the Raft Transport queue"}
> ```

// Replica queue metrics.

---

*[storage/metrics.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjBUN18AYRodR7zwKL:-KRjBUN18AYRodR7zwKM:b2m9mq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L155)):*
> ```Go
> 
> 	metaGCQueueProcessed = metric.Metadata{Name: "queue.gc.process.success",
> 		Help: "Total number of replicas processed by the GC queue"}
> ```

s/Total n/N/g

---

*[storage/metrics.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjBZx3jLqsm_jaB0Gd:-KRjBZx3jLqsm_jaB0Ge:b-arcpy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L160)):*
> ```Go
> 	metaGCQueuePending = metric.Metadata{Name: "queue.gc.pending",
> 		Help: "Number of pending replicas in the GC queue"}
> 	metaGCQueueTime = metric.Metadata{Name: "queue.gc.time",
> ```

throughout: i believe we use .nanos in the Name elsewhere - this is because it's a natural way to communicate to consumers what it is they're getting.

---

*[storage/metrics.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjC44z2JGcqkWNZLs4:-KRjC44z2JGcqkWNZLs5:b-t0uu0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L161)):*
> ```Go
> 		Help: "Number of pending replicas in the GC queue"}
> 	metaGCQueueTime = metric.Metadata{Name: "queue.gc.time",
> 		Help: "Nanoseconds spent processing replicas in the GC queue"}
> ```

i think we might want to use a gauge here - graphing per-replica time seems more useful than a running total.

---

*[storage/metrics.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjBqhFXGb0VVk7rEBW:-KRjBqhFXGb0VVk7rEBX:b-wthvqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L163)):*
> ```Go
> 		Help: "Nanoseconds spent processing replicas in the GC queue"}
> 	metaRaftLogQueueProcessed = metric.Metadata{Name: "queue.raftlog.process.success",
> 		Help: "Total number of replicas processed by the raft log queue"}
> ```

throughout: does this include failures? clarify.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

**Comment by spencerkimball** [comment]
Review status: 0 of 9 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[storage/gc_queue.go, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRirtwhG4IzF58X9wjr:-KRj-CjlyjwUY5Aqly75:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/gc_queue.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> to echo @petermattis from yesterday: this is Go, not obj-c, so I think we can omit the type from the name. Perhaps "processedReplicas"? ditto for the other fields.
</details>
OK

---

*[storage/metrics.go, line 153 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjCHtRUb02fuSQQmwy:-KRjEP3adCPuWywLEtcA:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L153)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> // Replica queue metrics.
</details>
OK

---

*[storage/metrics.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjBUN18AYRodR7zwKL:-KRjEVJVjHEEWYfykh6U:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L155)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Total n/N/g
</details>
OK

---

*[storage/metrics.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjBZx3jLqsm_jaB0Gd:-KRjEjwY13XLBeuRHSuo:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L160)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: i believe we use .nanos in the Name elsewhere - this is because it's a natural way to communicate to consumers what it is they're getting.
</details>
OK

---

*[storage/metrics.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjC44z2JGcqkWNZLs4:-KRjF3Rww9d02z8sBss4:b-qoztv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L161)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think we might want to use a gauge here - graphing per-replica time seems more useful than a running total.
</details>
Hmm, that's not how we measure time spent working on raft and ticking in raft.

---

*[storage/metrics.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjBqhFXGb0VVk7rEBW:-KRjExXlHWY7I76idHMm:b-tipma1) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L163)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: does this include failures? clarify.
</details>
No, clarified.

---

*[storage/metrics.go, line 426 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3T-r1-426:-KRixmS338Q4V284_M4u:bvvbr37) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L426)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The GCQueue could track more interesting information. See the `GCInfo` struct:
> 
> 
> 
> ```go
> 
> type GCInfo struct {
> 
> 	// Now is the timestamp used for age computations.
> 
> 	Now hlc.Timestamp
> 
> 	// Policy is the policy used for this garbage collection cycle.
> 
> 	Policy config.GCPolicy
> 
> 	// Stats about the userspace key-values considered, namely the number of
> 
> 	// keys with GC'able data, the number of "old" intents and the number of
> 
> 	// associated distinct transactions.
> 
> 	GCKeys, IntentsConsidered, IntentTxns int
> 
> 	// TransactionSpanTotal is the total number of entries in the transaction span.
> 
> 	TransactionSpanTotal int
> 
> 	// Summary of transactions which were found GCable (assuming that
> 
> 	// potentially necessary intent resolutions did not fail).
> 
> 	TransactionSpanGCAborted, TransactionSpanGCCommitted, TransactionSpanGCPending int
> 
> 	// AbortSpanTotal is the total number of transactions present in the abort cache.
> 
> 	AbortSpanTotal int
> 
> 	// AbortSpanConsidered is the number of abort cache entries old enough to be
> 
> 	// considered for removal. An "entry" corresponds to one transaction;
> 
> 	// more than one key-value pair may be associated with it.
> 
> 	AbortSpanConsidered int
> 
> 	// AbortSpanGCNum is the number of abort cache entries fit for removal (due
> 
> 	// to their transactions having terminated).
> 
> 	AbortSpanGCNum int
> 
> 	// PushTxn is the total number of pushes attempted in this cycle.
> 
> 	PushTxn int
> 
> 	// ResolveTotal is the total number of attempted intent resolutions in
> 
> 	// this cycle.
> 
> 	ResolveTotal int
> 
> 	// ResolveErrors is the number of successful intent resolutions.
> 
> 	ResolveSuccess int
> 
> 	// Threshold is the computed expiration timestamp. Equal to `Now - Policy`.
> 
> 	Threshold hlc.Timestamp
> 
> }
> 
> ```
</details>
This feels like another change. Care to file an issue?

---

*[storage/queue.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3L-r1-151:-KRiyhrKzzXajGMCO6Zb:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L151)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `s/count of total/total count of/g` throughout?
</details>
OK

---

*[storage/queue.go, line 526 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3L-r1-526:-KRiyrgeuoPgrB6cFtk6:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L526)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> also there's `log.ErrEvent{,f}` and we should probably get in the habit of using it whenever `err` or `pErr` is involved.
</details>
OK

---

*[storage/queue.go, line 528 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRisCwuTBsu02X9vxT_:-KRizKcHJvVq1Wbs7OBq:b4y7i3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L528)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: why are these nil checks necessary? if it's a matter of test construction, let's fix the tests.
</details>
`purgatoryMetric`, for one, is f always set. But OK, will get rid of the rest.

---

*[storage/queue.go, line 534 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3L-r1-534:-KRizIKnE44jnlj6_xIw:bppj29x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L534)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> wouldn't this make more sense next to `bq.failuresMetric`?
</details>
Not sure what you mean.

---

*[storage/queue.go, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3L-r1-548:-KRj-FCr9TqsUuNqWJhG:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L548)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Can we not make them never nil by creating dummy objects if none are provided?
</details>
OK

---

*[storage/queue_test.go, line 740 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiu1QrOB-1DL4AmRWJ:-KRjBzbfYGDuCweAa9Z4:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue_test.go#L740)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> all these arguments can be named `_`
</details>
OK

---

*[storage/queue_test.go, line 778 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3M-r1-778:-KRjCBFKc_kYuuWB2Uks:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue_test.go#L778)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit:
> 
> 
> 
> ```go
> 
> if min, v := 5*time.Millisecond, timeM.Count(); v < min.Nanoseconds() {
> 
>   return errors.Errorf("expected >= %s ns in processing time; got %d", min, v)
> 
> }
> 
> ```
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

File: storage/metrics.go (hunk 1 above)
Comment (tschottdorf) [inline storage/metrics.go]: It's really exactly the type of work you're doing here, but OK.

**Comment by tamird** [comment]
Reviewed 1 of 10 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/gc_queue.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjL6fLbpO6YxfS67kX:-KRjL6fLbpO6YxfS67kY:bl5hn44) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f952a8ace6bc0d73d012d1f97787d026d6035fd/storage/gc_queue.go#L108)):*
> ```Go
> 		failures:             store.metrics.GCQueueFailures,
> 		pending:              store.metrics.GCQueuePending,
> 		processingTime:       store.metrics.GCQueueTime,
> ```

s/Time/Nanos/g

that is also the local convention for existing metrics.

---

*[storage/metrics.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjC44z2JGcqkWNZLs4:-KRjLIGDercTfgOHt5IV:b-lblmae) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L161)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Hmm, that's not how we measure time spent working on raft and ticking in raft.
</details>
discussed with @petermattis, let's leave as is.

---

*[storage/queue.go, line 534 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3L-r1-534:-KRjMIBwizk62qZrE4XQ:bcvbrds) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L534)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Not sure what you mean.
</details>
I think @mberhault meant that there should be an error check next to the return from `processReplica`, and a call to `bq.successes.Inc(1)` and `bq.failures.Inc(1)` in its branches (rather than one here and one in the caller).

---

*[storage/queue.go, line 155 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjM0wIOCavoWrSWuzS:-KRjM0wIOCavoWrSWuzT:bk141oz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f952a8ace6bc0d73d012d1f97787d026d6035fd/storage/queue.go#L155)):*
> ```Go
> 	// failures is a total count of replicas which failed processing.
> 	failures *metric.Counter
> 	// pending is a gauge measuring current replica count pending.
> ```

below here you use "is a counter" or "is a gauge" but above you use "is a total count". i don't have a preference for either, but consistency would be good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

**Comment by tschottdorf** [comment]
Reviewed 1 of 10 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/metrics.go, line 426 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3T-r1-426:-KRjMzDGgoN0GRhGNuq1:b-9ug5nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/metrics.go#L426)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> It's really exactly the type of work you're doing here, but OK.
</details>
https://github.com/cockroachdb/cockroach/issues/9409

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

**Comment by spencerkimball** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/gc_queue.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjL6fLbpO6YxfS67kX:-KRjtQiUjX1H4sZUb5YM:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f952a8ace6bc0d73d012d1f97787d026d6035fd/storage/gc_queue.go#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Time/Nanos/g
> 
> that is also the local convention for existing metrics.
</details>
OK

---

*[storage/queue.go, line 534 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRiqRkDDFBNpsBxLT3L-r1-534:-KRjxHrjji4BYEYbxeOz:b-73ov6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f9bd206295540b7155231f04d7ef37cb51051b/storage/queue.go#L534)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think @mberhault meant that there should be an error check next to the return from `processReplica`, and a call to `bq.successes.Inc(1)` and `bq.failures.Inc(1)` in its branches (rather than one here and one in the caller).
</details>
there are multiple returns from `processReplica` so I didn't want code duplication at 3 call sites.

---

*[storage/queue.go, line 155 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjM0wIOCavoWrSWuzS:-KRjti7o_E5JUspdEsWg:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f952a8ace6bc0d73d012d1f97787d026d6035fd/storage/queue.go#L155)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> below here you use "is a counter" or "is a gauge" but above you use "is a total count". i don't have a preference for either, but consistency would be good.
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 10 files at r1, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/queue.go, line 543 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjySkDIKAdK5ays0pv:-KRjySkE63InzXVof_gx:b-liq8tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/eef41f59db8dbc3f27362ffe18b815ce81d116c9/storage/queue.go#L543)):*
> ```Go
> // purgatory.
> func (bq *baseQueue) maybeAddToPurgatory(repl *Replica, err error, clock *hlc.Clock, stopper *stop.Stopper) {
> 	// Increment failures metric.
> ```

surely this comment doesn't help

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-:-KRjyY-zbWsMkDndmtRV:bnfp4nl)*

**Comment by spencerkimball** [comment]
Review status: 3 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/queue.go, line 543 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398#-KRjySkDIKAdK5ays0pv:-KRk4sNJCVUFi8mle7QY:b56y6se) ([raw file](https://github.com/cockroachdb/cockroach/blob/eef41f59db8dbc3f27362ffe18b815ce81d116c9/storage/queue.go#L543)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> surely this comment doesn't help
</details>
Expanded

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9398)*

**Pull request merged**